heralding/capabilities/imap.py

```diff
@@ -0,0 +1,167 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import socket
+import base64
+import binascii
+
+from heralding.capabilities.handlerbase import HandlerBase
+
+logger = logging.getLogger(__name__)
+
+CRLF = '\r\n'
+
+
+class Imap(HandlerBase):
+    def __init__(self, options):
+        super(Imap, self).__init__(options)
+        self.max_tries = int(self.options['protocol_specific_data']['max_attempts'])
+        self.banner = self.options['protocol_specific_data']['banner']
+
+        self.available_commands = ['authenticate', 'capability', 'login', 'logout', 'noop']
+        self.available_mechanisms = ['plain']
+
+    def execute_capability(self, address, socket, session):
+        self._handle_session(session, socket)
+
+    def _handle_session(self, session, gsocket):
+        fileobj = gsocket.makefile()
+
+        self.send_message(session, gsocket, self.banner)
+
+        state = "Not Authenticated"
+        while state != 'Logout' and session.connected:
+            try:
+                raw_msg = fileobj.readline()
+            except socket.error:
+                session.end_session()
+                break
+
+            cmd_msg = raw_msg.rstrip().split(' ', 2)
+            if len(cmd_msg) == 0:
+                continue
+            elif len(cmd_msg) == 1:
+                self.send_message(session, gsocket, "* BAD invalid command")
+                continue
+            elif len(cmd_msg) == 2:
+                tag = cmd_msg[0]
+                cmd = cmd_msg[1]
+                args = ''
+            else:
+                tag = cmd_msg[0]
+                cmd = cmd_msg[1]
+                args = cmd_msg[2]
+
+            cmd = cmd.lower()
+            if cmd not in self.available_commands:
+                self.send_message(session, gsocket, tag + " BAD invalid command")
+            else:
+                func_to_call = getattr(self, 'cmd_{0}'.format(cmd), None)
+                if func_to_call:
+                    return_value = func_to_call(session, gsocket, tag, args)
+                    state = return_value
+                else:
+                    self.send_message(session, gsocket, tag + " BAD invalid command")
+        session.end_session()
+
+    def cmd_authenticate(self, session, gsocket, tag, args):
+        mechanism = args.split()
+        if len(mechanism) == 1:
+            auth_mechanism = mechanism[0].lower()
+        else:
+            self.send_message(session, gsocket, tag + ' BAD invalid command')
+            return "Not Authenticated"
```

Comment:
"Not Authenticated" value is used in comparison with "Logout" in while loop:

```
while state != 'Logout' and session.connected:
     try:
         raw_msg = fileobj.readline()
         except socket.error:
```
I decided that each method should return the string, which means current state of Imap session. It isn't important, but in my opinion it is simpler to understand, how each Imap command affects on the session state, while looking through the code.